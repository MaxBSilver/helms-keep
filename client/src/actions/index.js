export const storeQuests = quests => ({
  type: 'STORE_QUESTS',
  payload: {
    quests
  }
})

export const addQuest = quest => ({
  type: 'ADD_QUEST',
  payload: {
    quest
  }
})

export const editQuest = quest => {
  return {
    type: "EDIT_QUEST",
    payload: {
      quest
    }
  };
}

export const deleteQuest = id => {
  return {
    type: "DELETE_QUEST",
    payload: {
      id
    }
  };
}

export const showPopup = bool => {
  return {
    type: "SHOW_POPUP",
    payload: {
      bool
    }
  }
}

