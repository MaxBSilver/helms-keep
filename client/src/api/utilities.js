export const host = 'http://localhost:5000/';

export const mockTask = {
  id: 1234,
  isCompleted: false, 
  message: 'Clean the kitchen'
}

export const mockNote = {
  id: 1234,
  title: 'My First Quest',
  tasks: [mockTask, mockTask]
}