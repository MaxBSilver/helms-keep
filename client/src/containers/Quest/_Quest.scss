.Quest {
  margin: 0 35px 35px 0;
  border: 1px solid #999999;
  border-radius: 6px;
  box-shadow: 0 0 8px rgba(0, 0, 0, 0.639);
  height: 375px;
  width: 310px;
  position: relative;
  background: white;

  .green-quest {
    border-color: #8CAC81;
  }
}

.card-header {
  background: grey;
  padding: 12px 15px;
  border-radius: 4px 4px 0 0;
  display: flex;
  justify-content: space-between;

  .green-header {
    background: rgb(119, 148, 109);
  }

  button {
    background: none!important;
    border: none;
  }

  .expand-btn {
    font-size: 20px;
    color: white;
    opacity: .7;

    &:hover {
      opacity: 1;
    }
  }

  .coin {
    height: 20px;
    cursor: pointer;
  }
}

.card-title {
  font-size: 18px;
  font-weight: 600;
  color: white;
  cursor: text;
  flex: auto;
  padding-left: 12px;
}

.card-body {
  height: 290px;
  overflow-y: scroll;
  border-bottom: 1px solid #c9dacc;
  border-top: 1px solid #688f6f9c;
}

.delete-btn {
  background: none!important;
  border: none;
  font-size: 18px;
  color: $btnColor;

  &:hover {
    color: darkred!important;
    opacity: .7;
  }
}

.card-footer {
  background: lightgray;
  height: 40px;
  border-radius: 5px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding-bottom: 3px;

  .green-footer {
    background: #E1EBDE;
  }
}

.show-completed {
  font-size: 14px;
  color: $btnColor;

  &:hover {
    color: $btnHoverColor;
  }
}

.delete-btn,
.show-completed {
  cursor: pointer;
  margin: 0 15px;
}

.divider {
  opacity: .6;
  margin: 10px 0 5px 3px;
  font-weight: 500;
}
