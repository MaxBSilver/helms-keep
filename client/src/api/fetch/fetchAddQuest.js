import { host } from '../utilities';
import shortid from 'shortid';

export const fetchAddQuest = (title, challenges) => {
	const url = `${host}quests`;
	const body = { id: shortid.generate(), title, challenges };
	console.log(body);
	const options = {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(body)
	};
	return fetch(url, options).then(response => {
		if (!response.ok) {
			throw Error('Failed to add quest');
		} else {
			return response.json();
		}
	});
};
