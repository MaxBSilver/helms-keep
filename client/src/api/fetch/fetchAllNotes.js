import { host } from '../utilities';

export const fetchAllNotes = () => {
  return fetch(`${host}quests`)
    .then(response => {
      if (!response.ok) {
        throw Error('Failed to get quests')
      } else {
        return response.json();
      }
    })
}