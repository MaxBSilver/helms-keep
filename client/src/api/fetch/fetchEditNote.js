import { host } from '../utilities';

export const fetchEditNote = (note) => {
  const url = `${host}quests/${note.id}`;
  const options = {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(note)
  }
  return fetch(url, options)
    .then(response => {
      if (!response.ok) {
        throw Error('Failed to edit note')
      } else {
        return response.json();
      }
    })
}