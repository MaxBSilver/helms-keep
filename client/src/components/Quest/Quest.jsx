import React, { Component } from 'react';
import { fetchEditNote } from '../../api/fetch/fetchEditNote';
import { connect } from "react-redux";
import { editQuest, showPopup } from '../../actions';

export class Quest extends Component {
	constructor(props) {
		super(props);
		this.state = {
			showCompleted: false
		};
	}

	toggleShowCompleted = () => {
		this.setState({ showCompleted: !this.state.showCompleted });
  };

  // handleDelete = () => {
  //   const { id } = this.props.data;
  //   fetchDeleteNote(id);
  //   this.props.deleteQuest(id);
  // }
  
  handleTrashClick = () => {
    this.props.showPopup(true);
  }

  handleUpdate = (e) => {
    const localNote = {...this.props.data};
    const targetChallenge = localNote.challenges.find(chal => chal.id === e.target.id);
    if (e.target.classList[1] === "fa-square") {
      targetChallenge.isCompleted = true;
    } else if (e.target.className === "card-title" && e.key === "Enter") {
      e.preventDefault();
      localNote.title = e.target.innerText;
      e.target.blur();
    } else if (e.target.classList[1] === "fa-check-square") {
      targetChallenge.isCompleted = false;
    } else if (e.key === "Enter") {
      e.preventDefault();
      targetChallenge.message = e.target.innerText;
      e.target.blur();
    } else if (e.target.className === "message" && e.type === "blur") {
      targetChallenge.message = e.target.innerText;
    } else if (e.target.className === "card-title" && e.type === "blur") {
      localNote.title = e.target.innerText;
    }
    this.props.updateQuest(localNote);
    fetchEditNote(localNote);
  }

  render() {
    const { title, challenges } = this.props.data;
		const completedTaskItems = [];
    const uncompletedTaskItems = [];
    
    challenges.forEach(({ id, message, isCompleted }) => {
      let boxClass = isCompleted ? "fa-check-square" : "fa-square";
      let card = (
        <li className="challenge-txt" key={id}>
          <i
            className={`far ${boxClass}`}
            id={id}
            onClick={this.handleUpdate}
          />
          <span
            contentEditable="true"
            id={id}
            className='message'
            suppressContentEditableWarning={true}
            onKeyDown={this.handleUpdate}
            onBlur={this.handleUpdate}
          >
            {message}
          </span>
        </li>
      );
      isCompleted
        ? completedTaskItems.push(card)
        : uncompletedTaskItems.push(card);
    })

		return (
      <article className="Quest">
        <div className="card-header">
          <h2
            className="card-title"
            contentEditable="true"
            suppressContentEditableWarning={true}
            onKeyDown={this.handleUpdate}
            onBlur={this.handleUpdate}
          >
            {title}
          </h2>
        </div>
        <div className="card-body">
          <ul>{uncompletedTaskItems}</ul>
          <p className="show-completed" onClick={this.toggleShowCompleted}>
            Show completed ({completedTaskItems.length})
          </p>
          <ul>{this.state.showCompleted && completedTaskItems}</ul>
        </div>
        <button className="delete-btn" type="button">
          <i className="fas fa-trash" onClick={this.handleTrashClick}/>
        </button>
      </article>
    );
	}
}

export const mapDispatchToProps = dispatch => ({
  updateQuest: quest => dispatch(editQuest(quest)),
  showPopup: bool => dispatch(showPopup(bool))
})

export default connect(null, mapDispatchToProps)(Quest);
