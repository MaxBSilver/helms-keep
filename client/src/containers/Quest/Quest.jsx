import React, { Component } from 'react';
import { fetchEditNote } from '../../api/fetch/fetchEditNote';
import { connect } from "react-redux";
import { editQuest, showPopup } from '../../actions';
import ChallengeContainer from '../../components/ChallengeContainer/ChallengeContainer';

export class Quest extends Component {
	constructor(props) {
		super(props);
		this.state = {
			showCompleted: false
		};
	}

	toggleShowCompleted = () => {
		this.setState({ showCompleted: !this.state.showCompleted });
  };

  addChallenge = challenge => {
    const quest = { ...this.props.data };

    quest.challenges.push(challenge);
    this.updateQuest(quest)
  };

  deleteChallenge = id => {
    const quest = { ...this.props.data };

    quest.challenges.splice(quest.challenges.findIndex(chal => chal.id === id), 1);
    this.updateQuest(quest);
  }

  updateChallenge = challenge => {
    const quest = { ...this.props.data };
    const chalIndex = quest.challenges.findIndex(chal => chal.id === challenge.id);

    quest.challenges[chalIndex] = challenge;
    this.updateQuest(quest);
  }
  
  showDeleteWarning = () => {
    const { id } = this.props.data;
    this.props.showPopup(true, id, 'delete');
  }

  updateQuest = quest => {
    this.props.updateQuest(quest);
    fetchEditNote(quest);
  }

  updateTitle = e => {
    const quest = {...this.props.data};
    
    if (e.key === "Enter" || e.type === 'blur') {
      e.preventDefault();
      quest.title = e.target.innerText;
      e.target.blur();
    }
    this.updateQuest(quest);
  }

  render() {
    const { title, challenges, id } = this.props.data;
    const completedChallenges = challenges.filter(chal => chal.isCompleted);
    const incompleteChallenges = challenges.filter(chal => !chal.isCompleted);
    const verb = !this.state.showCompleted ? '+ Show' : '- Hide';

    const completed = this.state.showCompleted && completedChallenges.length ?    <>
        <h4 className="divider">Completed</h4>
        <ChallengeContainer
          type="complete"
          challenges={completedChallenges}
          addChallenge={this.addChallenge}
          updateChallenge={this.updateChallenge}
          deleteChallenge={this.deleteChallenge}
        />
      </>
      : null;

    const link = completedChallenges.length ?
      <p role="button" className="show-completed" onClick={this.toggleShowCompleted}>{verb} {completedChallenges.length} completed</p>
      : null;
    
		return (
      <article className="Quest">
        <div className="card-header">
          <h2
            className="card-title"
            contentEditable="true"
            suppressContentEditableWarning={true}
            onKeyDown={this.updateTitle}
            onBlur={this.updateTitle}
          >
            {title}
          </h2>
        </div>
        <div className="card-body">
          <ChallengeContainer
            type="incomplete"
            challenges={incompleteChallenges}
            addChallenge={this.addChallenge}
            updateChallenge={this.updateChallenge}
            deleteChallenge={this.deleteChallenge}
          />
          {completed}
        </div>
        <div className="card-footer">
          <div>{link}</div>
          <button className="delete-btn" type="button">
            <i className="fas fa-trash" onClick={this.showDeleteWarning}/>
          </button>
        </div>
      </article>
    );
	}
}

export const mapDispatchToProps = dispatch => ({
  updateQuest: quest => dispatch(editQuest(quest)),
  showPopup: (bool, id, type) => dispatch(showPopup(bool, id, type))
})

export default connect(null, mapDispatchToProps)(Quest);
