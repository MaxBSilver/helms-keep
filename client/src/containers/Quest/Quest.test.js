import React from 'react';
import { shallow } from 'enzyme';
import { Quest, mapDispatchToProps } from './Quest';
import { mockTask, mockCompletedTask, mockNote } from '../../api/utilities';
import { fetchEditNote } from '../../api/fetch/fetchEditNote';
import { editQuest, showPopup } from '../../actions';

// test

jest.mock('../../api/fetch/fetchEditNote');

describe('Quest', () => {
	let wrapper, instance;

	const mockShowPopUp = jest.fn();
	const mockUpdateQuest = jest.fn();
	const mockKeyEvent = {
		key: 'Enter',
		preventDefault: () => {},
		target: {
			innerText: 'New title',
			blur: jest.fn()
		}
	};
	const mockBlurEvent = {
		type: 'blur',
		preventDefault: () => {},
		target: {
			innerText: 'New title',
			blur: jest.fn()
		}
	}

	beforeEach(() => {
		const mockNote = {
			id: 1234,
			title: 'My First Quest',
			challenges: [mockTask, mockCompletedTask]
		}
		wrapper = shallow(
		<Quest 
			data={mockNote} 
			showPopup={mockShowPopUp}
			updateQuest={mockUpdateQuest}
		/>);
		instance = wrapper.instance();
	});

	it('should match snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('should have a default state', () => {
		const defaultState = { showCompleted: false };
		expect(wrapper.state()).toEqual(defaultState);
	});

	describe('Event Handlers', () => {
		it('should invoke toggleShowCompleted on click', () => {
			expect(wrapper.state('showCompleted')).toEqual(false);
			wrapper.find('.show-completed').simulate('click');
			expect(wrapper.state('showCompleted')).toEqual(true);
		});

		it('should invoke updateTitle when the title is edited', () => {
			jest.spyOn(instance, 'updateTitle');
			wrapper.find('.card-title').simulate('keydown', mockKeyEvent);
			expect(mockKeyEvent.target.blur).toHaveBeenCalled();
		});

		it('should invoke updateTitle when the title is blured', () => {
			jest.spyOn(instance, 'updateTitle');
			wrapper.find('.card-title').simulate('blur', mockBlurEvent);
			expect(mockKeyEvent.target.blur).toHaveBeenCalled();
		});

		it('should invoke showDeleteWarning when the delete button is clicked', () => {
			wrapper.find('.fa-trash').simulate('click');
			expect(mockShowPopUp).toHaveBeenCalledWith(true, 1234, 'delete')
		});
	});

	describe('toggleShowCompleted', () => {
		it('should toggle the showCompleted property in state', () => {
			expect(wrapper.state('showCompleted')).toEqual(false);
			instance.toggleShowCompleted();
			expect(wrapper.state('showCompleted')).toEqual(true);
			instance.toggleShowCompleted();
			expect(wrapper.state('showCompleted')).toEqual(false);
		});
	});

	describe('addChallenge', () => {
		it('should add a challenge to the challenges array of quest', () => {
			expect(instance.props.data.challenges).toHaveLength(2);
			instance.addChallenge(mockTask);
			expect(instance.props.data.challenges).toHaveLength(3);
		});

		it('should invoke updateQuest with the the new quest data', () => {
			jest.spyOn(instance, 'updateQuest');
			instance.addChallenge(mockTask);
			let localNote = { ...instance.props.data}
			localNote.challenges.push(mockTask);
			expect(instance.updateQuest).toHaveBeenCalledWith(localNote);
		});
	});

	describe('deleteChallenge', () => {
		it('should remove the appropriate challenge from the quest data', () => {
			expect(instance.props.data.challenges).toHaveLength(2);
			instance.deleteChallenge(5678);
			expect(instance.props.data.challenges).toHaveLength(1);
		});

		it('should invoke updateQuest with the the new quest data', () => {
			jest.spyOn(instance, 'updateQuest');
			instance.deleteChallenge(1234);
			let localNote = { ...instance.props.data }
			localNote.challenges.pop();
			expect(instance.updateQuest).toHaveBeenCalledWith(localNote);
		});
	});

	describe('updateChallenge', () => {
		let localChallenge = {
			id: 1234,
			isCompleted: false,
			message: 'update todo list'
		}
		it('should reassign the appropriate challenge', () => {
			instance.updateChallenge(localChallenge);
			expect(instance.props.data.challenges).toEqual([localChallenge, mockCompletedTask])
		});

		it('should invoke updateQuest with the update quest data', () => {
			jest.spyOn(instance, 'updateQuest');
			let localNote = { ...instance.props.data };
			localNote.challenges[0] = localChallenge;
			instance.updateChallenge(localChallenge);
			expect(instance.updateQuest).toHaveBeenCalledWith(localNote);
		});
	});

	describe('showDeleteWarning', () => {
		it('should invoke showPopup with the correct params', () => {
			instance.showDeleteWarning();
			expect(mockShowPopUp).toHaveBeenCalledWith(true, 1234, 'delete');
		});
	});

	describe('updateQuest', () => {
		it('should invoke the updateQuest dispatch with the correct params', () => {
			instance.updateQuest(mockNote);
			expect(mockUpdateQuest).toHaveBeenCalledWith(mockNote);
		});

		it('should invoke the fetchEditNote dispatch with the correct params', () => {
			instance.updateQuest(mockNote);
			expect(fetchEditNote).toHaveBeenCalledWith(mockNote);
		});
	});

	describe('updateTitle', () => {
		let updatedNote = { ...mockNote, title: 'New title'};

		it('should invoke updateQuest if the user presses enter', () => {
			jest.spyOn(instance, 'updateQuest');
			instance.updateTitle(mockKeyEvent);
			expect(instance.updateQuest).toHaveBeenCalledWith(updatedNote);
		});

		it('should invoke updateQuest if the title blurs', () => {
			jest.spyOn(instance, 'updateQuest');
			instance.updateTitle(mockBlurEvent);
			expect(instance.updateQuest).toHaveBeenCalledWith(updatedNote);
		});

		it('should not invoke updateQuest if the condition is not met', () => {
			jest.spyOn(instance, 'updateQuest');
			instance.updateTitle({});
			expect(instance.updateQuest).not.toHaveBeenCalledWith();
		});
	});

	describe('MDTP', () => {
		let mappedProps, mockDispatch;
		beforeEach(() => {
			mockDispatch = jest.fn();
			mappedProps = mapDispatchToProps(mockDispatch);
		});

		it('should call dispatch when updateQuest is called', () => {
			const actionToDispatch = editQuest(mockNote);
			mappedProps.updateQuest(mockNote);
			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
		});

		it('should call dispatch when showPopup is called', () => {
			const actionToDispatch = showPopup(true, 1, 'delete');
			mappedProps.showPopup(true, 1, 'delete');
			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
		});
	});
});
