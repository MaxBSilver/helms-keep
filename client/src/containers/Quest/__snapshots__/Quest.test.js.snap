// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quest should match snapshot 1`] = `undefined`;
